src/instrumentation/create-telemetry-middleware.ts

```diff
@@ -0,0 +1,321 @@
+ */
+
+import { getSerializedObjectSafe } from "./helpers/get-serialized-object-safe.js";
+import { createSpan } from "./helpers/create-span.js";
+import { IdNameManager } from "./helpers/id-name-manager.js";
+import { getErrorSafe } from "./helpers/get-error-safe.js";
+import { isDeepEqual, isEmpty } from "remeda";
+import type { BeeCallbacks } from "@/agents/bee/types.js";
+import type { InferCallbackValue } from "@/emitter/types.js";
+import type { GenerateCallbacks } from "@/llms/base.js";
+import { FrameworkError } from "@/errors.js";
+import { Version } from "@/version.js";
+import { Role } from "@/llms/primitives/message.js";
+import type { GetRunContext, RunInstance } from "@/context.js";
+import type { GeneratedResponse, FrameworkSpan } from "./types.js";
+import { activeTracesMap, buildTraceTree } from "./tracer.js";
+import { traceSerializer } from "./helpers/trace-serializer.js";
+import { INSTRUMENTATION_IGNORED_KEYS } from "./config.js";
+import { createFullPath } from "@/emitter/utils.js";
+import type { BeeAgent } from "@/agents/bee/agent.js";
+import { Serializable } from "@/internals/serializable.js";
+import { instrumentationLogger } from "./logger.js";
+
+export function createTelemetryMiddleware() {
+  return (context: GetRunContext<RunInstance, unknown>) => {
+    if (!context.emitter?.trace?.id) {
+      throw new FrameworkError(`Fatal error. Missing traceId`);
+    }
+
+    const traceId = context.emitter?.trace?.id;
+    if (activeTracesMap.has(traceId)) {
+      return;
+    }
+    activeTracesMap.set(traceId, context.instance.constructor.name);
+
+    instrumentationLogger.debug(
+      {
+        source: context.instance.constructor.name,
+        traceId: traceId,
+      },
+      "createTelemetryMiddleware",
+    );
+    const { emitter, runParams, instance } = context;
+    const basePath = createFullPath(emitter.namespace, "");
+
+    let prompt: string | undefined | null = null;
+    if (instance.constructor.name === "BeeAgent") {
+      prompt = (runParams as Parameters<BeeAgent["run"]>)[0].prompt;
+    }
+
+    const spansMap = new Map<string, FrameworkSpan>();
+    const parentIdsMap = new Map<string, number>();
+    const spansToDeleteMap = new Map<string, undefined>();
+
+    let generatedMessage: GeneratedResponse | undefined = undefined;
+    let history: GeneratedResponse[] | undefined = undefined;
+    const groupIterations: string[] = [];
+
+    const idNameManager = new IdNameManager();
+
+    const newTokenEventName: keyof GenerateCallbacks = `newToken`;
+    const partialUpdateEventName: keyof BeeCallbacks = "partialUpdate";
+    const successEventName: keyof GenerateCallbacks = `success`;
+    const finishEventName: keyof GenerateCallbacks = `finish`;
+    const startEventName: keyof GenerateCallbacks = `start`;
+    const errorEventName: keyof GenerateCallbacks = `error`;
+
+    const eventsIterationsMap = new Map<string, Map<string, string>>();
+
+    const startTime = new Date();
```

Comment:
Would not be `performance.now` better?

Reply:
Does not work with open telemetry framework. It would not help.